services/user/Tokens/src/RTokens.cpp

```diff
@@ -13,64 +15,229 @@ using namespace UserService;
+
+   struct TokenBalance
+   {
+      SID       symbolId;
+      TID       tokenId;
+      Precision precision;
+      uint64_t  balance;
+
+      auto getKey() const { return realBalance(balance, precision); }
+   };
+   PSIO_REFLECT(TokenBalance, symbolId, tokenId, precision, balance);
 
+   struct TokenOwnerDetails
+   {
+      AccountNumber account;
+      AccountNumber authService;
+   };
+   PSIO_REFLECT(TokenOwnerDetails, account, authService);
+
+   // Removed inflation details because the inflation params aren't yet supported
+   //   in the service.
+   struct TokenDetail
+   {
+      TID                id;
+      NID                ownerNft;
+      TokenOwnerDetails  ownerDetails;
+      psibase::Bitset<8> config;
+      Precision          precision;
+      SID                symbolId;
+      Quantity           currentSupply;
+      Quantity           maxSupply;
+   };
+   PSIO_REFLECT(TokenDetail,
+                id,
+                ownerNft,
+                ownerDetails,
+                config,
+                precision,
+                symbolId,
+                currentSupply,
+                maxSupply);
+
+}  // namespace TokenQueryTypes
+
+using namespace TokenQueryTypes;
 struct TokenQuery
 {
    auto allBalances() const
    {  //
-      return tokenService.index<BalanceTable, 0>();
+      return tokenService.open<BalanceTable>().getIndex<0>();
    }
-   auto userBalances(AccountNumber user) const
-   {  //
-      vector<UserBalanceRecord> balances;
-      for (auto tokenType : tokenService.index<TokenTable, 0>())
+
+   /// Provides a way to query all balances for a given user.
+   /// Further filtering can be done to restrict the results by balance.
+   auto userBalances(AccountNumber         user,
+                     optional<uint64_t>    balance_gt,
+                     optional<uint64_t>    balance_ge,
+                     optional<uint64_t>    balance_lt,
+                     optional<uint64_t>    balance_le,
+                     optional<uint32_t>    first,
+                     optional<uint32_t>    last,
+                     optional<std::string> before,
+                     optional<std::string> after) const
+   {
+      vector<TokenBalance> balances;
+
+      auto tokenTypeIdx = tokenService.open<TokenTable>().getIndex<0>();
+      auto balanceIdx   = tokenService.open<BalanceTable>().getIndex<1>().subindex(user);
+
+      // Balances of this user
+      for (auto balance : balanceIdx)
+      {
+         auto tokenOpt = tokenTypeIdx.get(balance.key.tokenId);
+         check(tokenOpt.has_value(), "Invalid token type");
+         balances.push_back(TokenBalance{tokenOpt->symbolId, balance.key.tokenId,
+                                         tokenOpt->precision, balance.balance});
+      }
+
+      return makeVirtualConnection<
+          Connection<TokenBalance, "UserBalanceConnection", "UserBalanceEdge">>(
+          balances, balance_gt, balance_ge, balance_lt, balance_le, first, last, before, after);
```

Comment:
`gt`, `lt`, etc., only make sense if the data is sorted by balance.

Reply:
Deleted those filters